microcosm_pubsub/result.py

```diff
@@ -0,0 +1,57 @@
+"""
+Message handling result.
+
+"""
+from dataclasses import dataclass
+from enum import Enum, unique
+
+from microcosm_pubsub.errors import Nack, SkipMessage
+
+
+@unique
+class MessageHandlingResultType(Enum):
+    # Messaging handling failed.
+    FAILED = "FAILED"
+
+    # Messaging handling was intentially retried.
+    RETRIED = "RETRIED"
+
+    # Message handling succeeded.
+    SUCCEEDED = "SUCCEEDED"
+
+    # Message handling was skipped.
+    # ("Upon closer inspection these are loafers")
+    SKIPPED = "SKIPPED"
+
+    def __str__(self):
+        return self.name
+
+
+@dataclass
+class MessageHandlingResult:
+    media_type: str
+    result: MessageHandlingResultType
+
+    @classmethod
+    def invoke(cls, func, message, **kwargs):
+        try:
+            success = func(message=message, **kwargs)
+            if success:
+                result = MessageHandlingResultType.SUCCEEDED
+            else:
+                result = MessageHandlingResultType.SKIPPED
+            message.ack()
+        except SkipMessage:
+            result = MessageHandlingResultType.SKIPPED
+            message.ack()
+        except Nack as nack:
+            result = MessageHandlingResultType.RETRIED
+            message.nack(nack.visibility_timeout_seconds)
+        except Exception as error:
+            result = MessageHandlingResultType.FAILED
```

Comment:
Technically speaking, this will also be retried (FAILED might indicate that this was a terminal state for this event).

Reply:
Better naming suggestions?

Reply:
How about: `DELAYED` vs `FAILED`?

Reply:
DELAYED sounds closer to the RETRIED case, actually.

If the message context is available here, we *could* still use FAILED if we knew the current retry count, and if we'd stop trying after this point.

Still, FAILED is probably more readable than something like UNHANDLED_ERROR, and it has a similar connotation anyways.

Reply:
Conclusion: no change here